File: learn/hashing/rolling-hash/rolling_hash.go


type RollingHash struct {
	base int64
	mod  int64
}

func NewRollingHash(base, mod int64) *RollingHash {
	return &RollingHash{base: base, mod: mod}
}

func (rh *RollingHash) ComputeHash(s string) int64 {
	var hashValue int64 = 0
	for _, ch := range s {
		hashValue = (hashValue*rh.base + int64(ch)) % rh.mod
	}
	return hashValue
}

func (rh *RollingHash) RabinKarpSearch(text, pattern string) []int {
	n := len(text)
	m := len(pattern)
	var result []int

	if m == 0 || m > n {
		return result
	}

	var patternHash int64 = 0
	var currentHash int64 = 0
	var h int64 = 1

	for i := 0; i < m-1; i++ {
		h = (h * rh.base) % rh.mod
	}

	for i := 0; i < m; i++ {
		patternHash = (patternHash*rh.base + int64(pattern[i])) % rh.mod
		currentHash = (currentHash*rh.base + int64(text[i])) % rh.mod
	}

	for i := 0; i <= n-m; i++ {
		if patternHash == currentHash {
			match := true
			for j := 0; j < m; j++ {
				if text[i+j] != pattern[j] {
					match = false
					break
				}
			}
			if match {
				result = append(result, i)
			}
		}
		if i < n-m {
			currentHash = (int64(text[i+m]) + rh.base*(currentHash-int64(text[i])*h)%rh.mod) % rh.mod
			if currentHash < 0 {
				currentHash += rh.mod
			}
		}
	}
	return result
}

File: learn/hashing/rolling-hash/rolling_hash.rs
pub struct RollingHash {
    base: i64,
    modulus: i64,
}

impl RollingHash {
    pub fn new(base: i64, modulus: i64) -> Self {
        RollingHash { base, modulus }
    }

    pub fn compute_hash(&self, s: &str) -> i64 {
        let mut hash_value: i64 = 0;
        for ch in s.chars() {
            hash_value = (hash_value * self.base + ch as i64) % self.modulus;
        }
        hash_value
    }

    pub fn rabin_karp_search(&self, text: &str, pattern: &str) -> Vec<usize> {
        let n = text.len();
        let m = pattern.len();
        let mut result = Vec::new();

        if m == 0 || m > n {
            return result;
        }

        let text_bytes = text.as_bytes();
        let pattern_bytes = pattern.as_bytes();

        let mut pattern_hash: i64 = 0;
        let mut current_hash: i64 = 0;
        let mut h: i64 = 1;

        for _ in 0..(m - 1) {
            h = (h * self.base) % self.modulus;
        }

        for i in 0..m {
            pattern_hash = (pattern_hash * self.base + pattern_bytes[i] as i64) % self.modulus;
            current_hash = (current_hash * self.base + text_bytes[i] as i64) % self.modulus;
        }

        for i in 0..=(n - m) {
            if pattern_hash == current_hash {
                let mut match_found = true;
                for j in 0..m {
                    if text_bytes[i + j] != pattern_bytes[j] {
                        match_found = false;
                        break;
                    }
                }
                if match_found {
                    result.push(i);
                }
            }
            if i < n - m {
                current_hash = (text_bytes[i + m] as i64
                    + self.base * (current_hash - (text_bytes[i] as i64) * h) % self.modulus)
                    % self.modulus;
                if current_hash < 0 {
                    current_hash += self.modulus;
                }
            }
        }
        result
    }
}
